lib/src/component/with_transition.dart

```diff
@@ -0,0 +1,410 @@
+    }
+
+    // getDerivedStateFromProps
+    if (props.isShown && _isOrWillBeHidden(transitionPhase.value)) {
+      _handleShow(props, transitionPhase.value);
+    } else if (!props.isShown && _isOrWillBeShown(transitionPhase.value)) {
+      _handleHide(props, transitionPhase.value);
+    }
+
+    // ----- transitionend event handling ----- //
+
+    /// Whether the overlay is not guaranteed to transition in response to the current
+    /// state change.
+    ///
+    /// _Stored as variable as workaround for not adding breaking change to [handleHiding] API._
+    ///
+    /// A transition may not always occur when the state moves from SHOWING to HIDING
+    /// if the PRE_SHOWING-->SHOWING-->HIDING transition happens back-to-back.
+    ///
+    /// Better to not always transition when the user is ninja-toggling a transitionable
+    /// component than to break state changes waiting for a transition that will never happen.
+    final _transitionNotGuaranteed = useRef(false);
+
+    /// Timer used to determine if a transition timeout has occurred.
+    final _transitionEndTimer = useRef<Timer>();
+
+    /// Stream for listening to `transitionend` events on the node.
+    final _endTransitionSubscription = useRef<StreamSubscription>();
+
+    void _cancelTransitionEventListener() {
+      _endTransitionSubscription.current?.cancel();
+      _endTransitionSubscription.current = null;
+    }
+
+    void _cancelTransitionEndTimer() {
+      _transitionEndTimer.current?.cancel();
+      _transitionEndTimer.current = null;
+    }
+
+    /// Listens for the next `transitionend` event and invokes a callback after
+    /// the event is dispatched.
+    void onNextTransitionEnd(Function() complete) {
+      var transitionCount = _isOrWillBeHidden(transitionPhase.value) ? _transitionOutCount(props) : _transitionInCount(props);
+
+      _cancelTransitionEventListener();
+      _cancelTransitionEndTimer();
+
+      _transitionEndTimer.current = Timer(props.transitionTimeout, () {
+        assert(ValidationUtil.warn(
+            'The number of transitions expected to complete have not completed. Something is most likely wrong.',
+        ));
+
+        _cancelTransitionEventListener();
+
+        complete();
+      });
+
+      _endTransitionSubscription.current =
+          transitionNodeRef.current?.onTransitionEnd?.skip(transitionCount - 1)?.take(1)?.listen((_) {
+        _cancelTransitionEndTimer();
+
+        complete();
+      });
+    }
+
+    // ----- Effects (Lifecycle) ----- //
+
+    final _isMounted = useRef(false);
+
+    // Mount + Unmount
+    useEffect(() {
+      // componentDidMount
+      _isMounted.current = true;
+
+      // componentWillUnmount
+      return () {
+        _isMounted.current = false;
+
+        _cancelTransitionEndTimer();
+        _cancelTransitionEventListener();
+      };
+    }, const []);
+
+    // componentDidUpdate
+    useLayoutEffect(() {
+      if (!_isMounted.current) return;
```

Comment:
Will this ever get called after the component is unmounted?

Reply:
No, but it does get called once before the component mounts.